File: app/src/main/java/com/jcons/pdfconverter/pdfconverter/IndirectObject.java
package com.jcons.pdfconverter.pdfconverter;


import java.io.IOException;
import java.io.OutputStream;

public class IndirectObject extends Base {

    private EnclosedContent mContent;
    private Dictionary mDictionaryContent;
    private Stream mStreamContent;
    private IndirectIdentifier mID;
    private int mByteOffset;
    private boolean mInUse;

    public IndirectObject() {
        clear();
    }

    public void setNumberID(int value) {
        mID.setNumber(value);
    }

    public int getNumberID() {
        return mID.getNumber();
    }

    public void setGeneration(int value) {
        mID.setGeneration(value);
    }

    public int getGeneration() {
        return mID.getGeneration();
    }

    public String getIndirectReference() {
        return mID.toPDFString() + " R";
    }

    public void setByteOffset(int value) {
        mByteOffset = value;
    }

    public int getByteOffset() {
        return mByteOffset;
    }

    public void setInUse(boolean value) {
        mInUse = value;
    }

    public boolean getInUse() {
        return mInUse;
    }


    public void setContent(String value) {
        mContent.setContent(value);
    }

    public void addDictionaryContent(String value) {
        mDictionaryContent.addContent(value);
    }

    public void setDictionaryContent(String value) {
        mDictionaryContent.setContent(value);
    }

    public void addStreamContent(String value) {
        mStreamContent.addContent(value);
    }

    public void setStreamContent(String value) {
        mStreamContent.setContent(value);
    }

    public String getStreamContent() {
        return mStreamContent.getContent();
    }

    protected String render() {
        StringBuilder sb = new StringBuilder();
        sb.append(mID.toPDFString());
        sb.append(" ");
        // j-a-s-d: this can be performed in inherited classes DictionaryObject and StreamObject
        if (mDictionaryContent.hasContent()) {
            mContent.setContent(mDictionaryContent.toPDFString());
            if (mStreamContent.hasContent())
                mContent.addContent(mStreamContent.toPDFString());
        }
        sb.append(mContent.toPDFString());
        return sb.toString();
    }

    @Override
    public void clear() {
        mID = new IndirectIdentifier();
        mByteOffset = 0;
        mInUse = false;
        mContent = new EnclosedContent("obj\n", "endobj\n");
        mDictionaryContent = new Dictionary();
        mStreamContent = new Stream();
    }

    @Override
    public String toPDFString() {
        return render();
    }

    public int writeTo(OutputStream stream) throws IOException {
        // j-a-s-d: this can be performed in inherited classes DictionaryObject and StreamObject
        if (mDictionaryContent.hasContent()) {
            mContent.setContent(mDictionaryContent.toPDFString());
            if (mStreamContent.hasContent())
                mContent.addContent(mStreamContent.toPDFString());
        }

        int offset = mID.writeTo(stream);

        stream.write(' ');
        offset++;

        return offset + mContent.writeTo(stream);
    }
}

File: app/src/main/java/com/jcons/pdfconverter/pdfconverter/IndirectObject.kt
package com.jcons.pdfconverter.pdfconverter

import java.io.IOException
import java.io.OutputStream


class IndirectObject : Base {
    private var mContent: EnclosedContent? = null
    private var mDictionaryContent: Dictionary? = null
    private var mStreamContent: Stream? = null
    private var mID: IndirectIdentifier? = null
    var byteOffset = 0
    var inUse = false

    var numberID: Int
        get() = mID!!.number
        set(value) {
            mID!!.number = value
        }

    var generation: Int
        get() = mID!!.generation
        set(value) {
            mID!!.generation = value
        }

    val indirectReference: String
        get() = mID!!.toPDFString() + " R"

    fun setContent(value: String?) {
        mContent?.content = value
    }

    fun addDictionaryContent(value: String?) {
        mDictionaryContent!!.addContent(value)
    }

    fun setDictionaryContent(value: String?) {
        mDictionaryContent?.content = value
    }

    fun addStreamContent(value: String?) {
        mStreamContent!!.addContent(value)
    }

    var streamContent: String?
        get() = mStreamContent?.content
        set(value) {
            mStreamContent?.content = value
        }

    protected fun render(): String {
        val sb = StringBuilder()
        sb.append(mID!!.toPDFString())
        sb.append(" ")
        // j-a-s-d: this can be performed in inherited classes DictionaryObject and StreamObject
        if (mDictionaryContent!!.hasContent()) {
            mContent?.content = mDictionaryContent!!.toPDFString()
            if (mStreamContent!!.hasContent()) mContent!!.addContent(mStreamContent!!.toPDFString())
        }
        sb.append(mContent!!.toPDFString())
        return sb.toString()
    }

    override fun clear() {
        mID = IndirectIdentifier()
        byteOffset = 0
        inUse = false
        mContent = EnclosedContent("obj\n", "endobj\n")
        mDictionaryContent = Dictionary()
        mStreamContent = Stream()
    }

    override fun toPDFString(): String {
        return render()
    }

    @Throws(IOException::class)
    fun writeTo(stream: OutputStream): Int {
        // j-a-s-d: this can be performed in inherited classes DictionaryObject and StreamObject
        if (mDictionaryContent!!.hasContent()) {
            mContent?.content = mDictionaryContent!!.toPDFString()
            if (mStreamContent!!.hasContent()) mContent!!.addContent(mStreamContent!!.toPDFString())
        }
        var offset = mID!!.writeTo(stream)
        stream.write(' '.toInt())
        offset++
        return offset + mContent!!.writeTo(stream)
    }

    init {
        clear()
    }
}
